File: android/app/src/test/kotlin/com/arogyax/signal/GoldenVectorsTest.kt
package com.arogyax.signal

import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File

/**
 * Checks this Kotlin port against the exact same golden vectors
 * app/test/fixtures/golden_vectors.json was written for - the file
 * app/test/policy_test.dart and friends were meant to assert against before
 * ticket 019 moved the app target to Kotlin. Same fixtures, same tolerance
 * (1e-6, per CLAUDE.md), new language.
 *
 * If this fails, either the Kotlin port has a bug or a DSP constant drifted
 * out of sync with the Dart sources under app/lib/signal and
 * ml/reference/dsp_reference.py - regenerate the fixture via
 * `python ml/reference/validate_dsp.py` and check which side is actually
 * right before touching this test.
 */
class GoldenVectorsTest {
    private val fixture: JSONObject by lazy {
        // Gradle unit tests run with the module dir (android/app) as the
        // working directory - repo root is two levels up.
        val file = File("../../app/test/fixtures/golden_vectors.json")
        require(file.exists()) { "golden_vectors.json not found at ${file.absolutePath}" }
        JSONObject(file.readText())
    }

    private val fs: Double by lazy { fixture.getDouble("fs") }

    private fun eachCase(block: (name: String, samples: DoubleArray, case: JSONObject) -> Unit) {
        val cases = fixture.getJSONArray("cases")
        for (i in 0 until cases.length()) {
            val case = cases.getJSONObject(i)
            val samplesJson = case.getJSONArray("samples")
            val samples = DoubleArray(samplesJson.length()) { samplesJson.getDouble(it) }
            block(case.getString("name"), samples, case)
        }
    }

    @Test
    fun `SQI matches the golden vectors`() {
        eachCase { name, samples, case ->
            val expected = case.getJSONObject("expectedSqi")
            val actual = SqiAnalyser(fs).analyse(samples)

            assertEquals("$name: sqi.score", expected.getDouble("score"), actual.score, 1e-6)
            assertEquals(
                "$name: sqi.saturationFraction",
                expected.getDouble("saturationFraction"),
                actual.saturationFraction,
                1e-6,
            )
            assertEquals(
                "$name: sqi.flatlineFraction",
                expected.getDouble("flatlineFraction"),
                actual.flatlineFraction,
                1e-6,
            )
            assertEquals(
                "$name: sqi.powerlineRatio",
                expected.getDouble("powerlineRatio"),
                actual.powerlineRatio,
                1e-6,
            )
            assertEquals(
                "$name: sqi.baselineWanderRatio",
                expected.getDouble("baselineWanderRatio"),
                actual.baselineWanderRatio,
                1e-6,
            )
        }
    }

    @Test
    fun `R-peak detection matches the golden vectors`() {
        eachCase { name, samples, case ->
            val expectedPeaksJson = case.getJSONArray("expectedPeaks")
            val expectedPeaks = List(expectedPeaksJson.length()) { expectedPeaksJson.getInt(it) }

            val actualPeaks = PanTompkins(fs).detect(samples).peaks

            assertEquals("$name: peak count", expectedPeaks.size, actualPeaks.size)
            assertEquals("$name: peak indices", expectedPeaks, actualPeaks)
        }
    }

    @Test
    fun `RR features match the golden vectors`() {
        eachCase { name, samples, case ->
            val result = PanTompkins(fs).detect(samples)
            val rr = RrAnalyser().analyse(result.rrIntervalsMs(fs))
            val expected = case.getJSONObject("expectedRrFeatures")

            assertEquals("$name: rr.count", expected.getInt("count"), rr.count)
            assertEquals("$name: rr.meanRrMs", expected.getDouble("meanRrMs"), rr.meanRrMs, 1e-6)
            assertEquals("$name: rr.meanHr", expected.getDouble("meanHr"), rr.meanHr, 1e-6)
            assertEquals("$name: rr.rmssdMs", expected.getDouble("rmssdMs"), rr.rmssdMs, 1e-6)
            assertEquals(
                "$name: rr.normalisedRmssd",
                expected.getDouble("normalisedRmssd"),
                rr.normalisedRmssd,
                1e-6,
            )
            assertEquals("$name: rr.pnn50", expected.getDouble("pnn50"), rr.pnn50, 1e-6)
            assertEquals(
                "$name: rr.normalisedShannonEntropy",
                expected.getDouble("normalisedShannonEntropy"),
                rr.normalisedShannonEntropy,
                1e-6,
            )
            assertEquals(
                "$name: rr.irregularityScore",
                expected.getDouble("irregularityScore"),
                rr.irregularityScore,
                1e-6,
            )
            assertEquals(
                "$name: rr.rejectedIntervals",
                expected.getInt("rejectedIntervals"),
                rr.rejectedIntervals,
            )
        }
    }

    @Test
    fun `fixture actually has cases - a green suite with zero cases proves nothing`() {
        assertTrue(fixture.getJSONArray("cases").length() > 0)
    }
}

File: app/test/golden_vectors_test.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:flutter_test/flutter_test.dart';

import 'package:arogyax/signal/pan_tompkins.dart';
import 'package:arogyax/signal/rr_features.dart';
import 'package:arogyax/signal/sqi.dart';

/// Pins the Dart DSP to the verified Python reference implementation.
///
/// The vectors in test/fixtures/golden_vectors.json are produced by
/// ml/reference/validate_dsp.py, whose filter design is checked against scipy
/// and whose R-peak detection is measured against synthetic signals with exact
/// known peak positions. If these tests fail, the two implementations have
/// drifted apart, and the Python side is the one that has been independently
/// verified.
///
/// Regenerate with:  python ml/reference/validate_dsp.py
void main() {
  late Map<String, dynamic> vectors;
  late double fs;

  setUpAll(() {
    final file = File('test/fixtures/golden_vectors.json');
    if (!file.existsSync()) {
      fail('Golden vectors missing. Run: python ml/reference/validate_dsp.py');
    }
    vectors = json.decode(file.readAsStringSync()) as Map<String, dynamic>;
    fs = (vectors['fs'] as num).toDouble();
  });

  test('fixture sample rate matches the BLE contract', () {
    expect(fs, 250.0, reason: 'contracts/ble.md fixes the sample rate at 250 Hz');
  });

  group('golden vectors', () {
    for (final caseName in const [
      'nsr_clean',
      'af_clean',
      'nsr_mains',
      'af_noisy',
    ]) {
      Map<String, dynamic> caseData() => (vectors['cases'] as List)
          .cast<Map<String, dynamic>>()
          .firstWhere((c) => c['name'] == caseName);

      Float64List samples() => Float64List.fromList(
            (caseData()['samples'] as List).map((v) => (v as num).toDouble()).toList(),
          );

      group(caseName, () {
        test('R-peak detection recovers the ground-truth beats', () {
          final c = caseData();
          final truth = (c['groundTruthPeaks'] as List).cast<int>();
          final detected = PanTompkins(fs).detect(samples()).peaks;

          // Match within 50 ms, then require both sensitivity and positive
          // predictive value at 95%. Counting peaks alone would let a detector
          // that finds the wrong beats in the right number pass.
          final tolerance = (0.050 * fs).round();
          final unmatched = List<int>.from(truth);
          var tp = 0;
          for (final d in detected) {
            var best = -1, bestDist = tolerance + 1;
            for (var i = 0; i < unmatched.length; i++) {
              final dist = (d - unmatched[i]).abs();
              if (dist < bestDist) {
                best = i;
                bestDist = dist;
              }
            }
            if (best >= 0) {
              unmatched.removeAt(best);
              tp++;
            }
          }
          final sensitivity = tp / truth.length;
          final ppv = detected.isEmpty ? 0.0 : tp / detected.length;

          expect(sensitivity, greaterThanOrEqualTo(0.95),
              reason: 'missed beats fabricate long RR intervals');
          expect(ppv, greaterThanOrEqualTo(0.95),
              reason: 'spurious beats fabricate short RR intervals, which mimic AF');
        });

        test('RR features match the reference implementation', () {
          final c = caseData();
          final expected = c['expectedRrFeatures'] as Map<String, dynamic>;
          final peaks = PanTompkins(fs).detect(samples());
          final actual = const RrAnalyser().analyse(peaks.rrIntervalsMs(fs));

          expect(actual.count, expected['count'],
              reason: 'interval count must match exactly');
          for (final field in const [
            'meanRrMs',
            'meanHr',
            'rmssdMs',
            'normalisedRmssd',
            'pnn50',
            'normalisedShannonEntropy',
            'irregularityScore',
          ]) {
            expect(actual.toJson()[field] as double,
                closeTo((expected[field] as num).toDouble(), 1e-6),
                reason: '$field diverged from the Python reference');
          }
        });

        test('SQI matches the reference implementation', () {
          final c = caseData();
          final expected = c['expectedSqi'] as Map<String, dynamic>;
          final actual = SqiAnalyser(fs).analyse(samples());

          for (final field in const [
            'score',
            'saturationFraction',
            'flatlineFraction',
            'powerlineRatio',
            'baselineWanderRatio',
          ]) {
            expect(actual.toJson()[field] as double,
                closeTo((expected[field] as num).toDouble(), 1e-6),
                reason: '$field diverged from the Python reference');
          }
        });

        test('irregularity score lands on the correct side of the gate', () {
          final c = caseData();
          final isAf = c['expectedRhythm'] == 'af';
          final peaks = PanTompkins(fs).detect(samples());
          final f = const RrAnalyser().analyse(peaks.rrIntervalsMs(fs));

          if (isAf) {
            expect(f.irregularityScore, greaterThanOrEqualTo(0.5),
                reason: 'simulated AF must clear the irregularity gate');
          } else {
            expect(f.irregularityScore, lessThan(0.5),
                reason: 'simulated sinus rhythm must not clear the gate');
          }
        });
      });
    }
  });
}
